src/gt4py/next/ffront/foast_passes/type_deduction.py

```diff
@@ -835,9 +837,18 @@ def _visit_astype(self, node: foast.Call, **kwargs) -> foast.Call:
                 f"Invalid call to `astype`. Second argument must be a scalar type, but got {new_type}.",
             )
 
-        return_type = with_altered_scalar_kind(
-            value.type, getattr(ts.ScalarKind, new_type.id.upper())
-        )
+        if isinstance(value, foast.TupleExpr):
+            return_type = type_info.apply_to_primitive_constituents(
+                value.type,
+                lambda primitive_type: with_altered_scalar_kind(
+                    primitive_type, getattr(ts.ScalarKind, new_type.id.upper())
+                ),
+            )
+
+        else:
+            return_type = with_altered_scalar_kind(
+                value.type, getattr(ts.ScalarKind, new_type.id.upper())
+            )
```

Comment:
```suggestion
        return_type = type_info.apply_to_primitive_constituents(
                value.type,
                lambda primitive_type: with_altered_scalar_kind(
                    primitive_type, getattr(ts.ScalarKind, new_type.id.upper())
                ),
            )
```
Almost there :-P. `apply_to_primitive_constituents` also works on a scalar, that way the `with_altered_scalar_kind` call is not duplicated.